# Imports

https://data.cityofchicago.org/Transportation/Traffic-Crashes-Crashes/85ca-t3if

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns

pd.set_option('display.max_columns', None)

# Functions

## Clean column names

In [3]:
def clean_col_names(df, char=None):
    if char:
        for ch in char:
            df.columns = df.columns.str.replace(ch, '').str.lower()
            return df.columns
    else:
        df.columns = df.columns.str.lower()
        return df.columns


## missing info

In [4]:
def missing_info(df):
    missing_info_cols = []
    for col in df.columns:
        if df[col].isna().sum() > 0:
            missing_info_cols.append(col)
        else:
            continue
    for cl in missing_info_cols:
        print(cl, df[cl].value_counts(dropna=False), "\n\n")

In [5]:
def remove_unknowns(data):
    categorical_feature_mask = data.dtypes==object
    categorical_cols = data.columns[categorical_feature_mask].tolist()
    categorical_cols.remove('prim_contributory_cause')

    mult_unique = []
    for col in categorical_cols:
        if data[col].nunique() > 2:
            mult_unique.append(col)
        else:
            continue

    unknown = []
    for col in mult_unique:
        if (data[col] == "UNKNOWN").any():
            unknown.append(col)


    indexes = []
    for col in unknown:
        unknows = data.loc[data[col] == "UNKNOWN"].index
        indexes.extend(unknows)
        np.unique(indexes)
    data.drop(index=indexes, inplace=True)



# Data Cleaning


## Loading Datas (changes)

In [6]:
df = pd.read_csv("Traffic_Crashes_-_Crashes.csv")
clean_col_names(df, ["_I"])
df.rename(columns= {'crash_type':"injury/tow_or_no", 'first_crash_type':"crash_type"}, inplace=True)

In [7]:
nodet = df.loc[df["prim_contributory_cause"] == "UNABLE TO DETERMINE"].index
not_determined_df = df.loc[df["prim_contributory_cause"] == "UNABLE TO DETERMINE"]

In [8]:
df.drop(index=nodet, inplace=True)

In [9]:
remove_unknowns(df)
remove_unknowns(not_determined_df)

/Users/Alisa/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/frame.py:4117: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [10]:
for col in list(df.columns):
    col.value_counts()

AttributeError: 'str' object has no attribute 'value_counts'

In [22]:
df == 'OTHER'

crash_recordd  rd_no  crash_date_est  crash_date  posted_speed_limit  \
0               False  False           False       False               False   
4               False  False           False       False               False   
10              False  False           False       False               False   
13              False  False           False       False               False   
14              False  False           False       False               False   
...               ...    ...             ...         ...                 ...   
395195          False  False           False       False               False   
395198          False  False           False       False               False   
395200          False  False           False       False               False   
395202          False  False           False       False               False   
395203          False  False           False       False               False   

        traffic_control_device  device_condition  weather_condition  \
0                        False             False              False   
4                        False             False              False   
10                       False             False              False   
13                       False             False              False   
14                       False             False              False   
...                        ...               ...                ...   
395195                   False             False              False   
395198                   False             False              False   
395200                   False             False              False   
395202                   False             False              False   
395203                   False             False              False   

        lighting_condition  crash_type  trafficway_type  lane_cnt  alignment  \
0                    False       False            False     False      False   
4                    False       False            False     False      False   
10                   False       False            False     False      False   
13                   False       False            False     False      False   
14                   False       False            False     False      False   
...                    ...         ...              ...       ...        ...   
395195               False       False            False     False      False   
395198               False       False            False     False      False   
395200               False       False            False     False      False   
395202               False       False            False     False      False   
395203               False       False            False     False      False   

        roadway_surface_cond  road_defect  report_type  injury/tow_or_no  \
0                      False        False        False             False   
4                      False        False        False             False   
10                     False        False        False             False   
13                     False        False        False             False   
14                     False        False        False             False   
...                      ...          ...          ...               ...   
395195                 False        False        False             False   
395198                 False        False        False             False   
395200                 False        False        False             False   
395202                 False        False        False             False   
395203                 False        False        False             False   

        intersection_related  not_right_of_way  hit_and_run  damage  \
0                      False             False        False   False   
4                      False             False        False   False   
10                     False             False        False   False   
13                     False        

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 252089 entries, 0 to 395203
Data columns (total 49 columns):
crash_recordd                    252089 non-null object
rd_no                            249562 non-null object
crash_date_est                   14208 non-null object
crash_date                       252089 non-null object
posted_speed_limit               252089 non-null int64
traffic_control_device           252089 non-null object
device_condition                 252089 non-null object
weather_condition                252089 non-null object
lighting_condition               252089 non-null object
crash_type                       252089 non-null object
trafficway_type                  252089 non-null object
lane_cnt                         130288 non-null float64
alignment                        252089 non-null object
roadway_surface_cond             252089 non-null object
road_defect                      252089 non-null object
report_type                      246043 non-null o

In [8]:
not_determined_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143115 entries, 1 to 395201
Data columns (total 49 columns):
crash_recordd                    143115 non-null object
rd_no                            141541 non-null object
crash_date_est                   15052 non-null object
crash_date                       143115 non-null object
posted_speed_limit               143115 non-null int64
traffic_control_device           143115 non-null object
device_condition                 143115 non-null object
weather_condition                143115 non-null object
lighting_condition               143115 non-null object
crash_type                       143115 non-null object
trafficway_type                  143115 non-null object
lane_cnt                         68263 non-null float64
alignment                        143115 non-null object
roadway_surface_cond             143115 non-null object
road_defect                      143115 non-null object
report_type                      139863 non-null ob

Everything that I do to **df** i will do to **not_determined_df** 

## CHECK for missing data or UNKNOWN


In [23]:
missing_info_cols = []
for col in df.columns:
    if df[col].isna().sum() > 0:
        missing_info_cols.append(col)
    else:
        continue

In [24]:
for col in missing_info_cols:
    print(col, "\n",  df[col].isna().sum(), "\n")

rd_no 
 2148 

crash_date_est 
 202485 

lane_cnt 
 101722 

report_type 
 5109 

intersection_related 
 156353 

not_right_of_way 
 204154 

hit_and_run 
 169220 

street_direction 
 1 

beat_of_occurrence 
 2 

photos_taken 
 210425 

statements_taken 
 208208 

dooring 
 212780 

work_zone 
 211825 

work_zone_type 
 212166 

workers_present 
 213013 

num_units 
 880 

most_severenjury 
 1188 

injuries_total 
 1182 

injuries_fatal 
 1182 

injuriesncapacitating 
 1182 

injuries_nonncapacitating 
 1182 

injuries_reported_not_evident 
 1182 

injuries_nondication 
 1182 

injuries_unknown 
 1182 

latitude 
 1365 

longitude 
 1365 

location 
 1365 



In [25]:
df['prim_contributory_cause'].value_counts(dropna=False)

FAILING TO YIELD RIGHT-OF-WAY                                                       38833
FOLLOWING TOO CLOSELY                                                               37294
NOT APPLICABLE                                                                      16643
IMPROPER OVERTAKING/PASSING                                                         15933
IMPROPER BACKING                                                                    14813
FAILING TO REDUCE SPEED TO AVOID CRASH                                              13996
IMPROPER LANE USAGE                                                                 13863
IMPROPER TURNING/NO SIGNAL                                                          11474
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  9867
DISREGARDING TRAFFIC SIGNALS                                                         5689
WEATHER                                                                              5471
OPERATING 

There are 143,115 unable to determine

- get rid of them?

Got rid of these above -- will use in future


Drop (not helpful)

- RD_NO - alot missing; not helpful for model (unique value)
- LANE_CNT - 19,000 missing; there are really big #; WHERE DID THIS EVEN HAPPEN?
- CRASH_RECORD_ID
- crash_date_est


Y Or N

- CRASH_DATE_EST_I - alot of missing values(365944); Crash date estimated by desk officer or reporting party (only used in cases where crash is reported at police station days after the crash)
- REPORT_TYPE - 9298 missing; **CAN BE USED instead of THE ABOVE** ; Administrative report type (at scene, at desk, amended)


- INTERSECTION_RELATED_I - alot missing;  **how accident happened**
- NOT_RIGHT_OF_WAY_I - alot missing;   **how accident happened**
- HIT_AND_RUN_I - alot missing;  **how accident happened**
- DOORING - alot missing; **how accident happened**


- WORK_ZONE_I - alot missing; **if accident happend on workzone**
- WORKERS_PRESENT_I - ALOT missing; 


- PHOTOS TAKEN - ALOT missing; do we really need this column?
- STATEMENTS_TAKEN_I - alot missing; do we really need this column?


EASY TO DEAL WITH

- STREET_DIRECTION - only 2 missing values; DROP MISING VALUES last
- STREET_NAME - only one missing; DROP MISSING VALUE last
- BEAT_OF_OCCURANCE - only 4 missing; DROP MISSING VALUE last (the different sections of chicago)


- NUM_UNITs - only 1,597 missing; MAYBE DROP MISSING VALUES LAST (# people in accident)


- MOST_SEVERE_INJURY - only 2,421 missing DROP AT END
- INJURIES_TOTAL- only 2414 missing; drop at end
- INJURIES_FATAL - only 2414 missing; drop at end
- INJURIES_INCAPACITATING - 
- INJURIES_NON_INCAPACITATING -
- INJURIES_REPORTED_NOT_EVIDENT - 
- INJURIES_NO_INDICATION - 
- INJURIES_UNKNOWN - 


- LATITUDE - only 2,160 missing look at end; maybe drop
- LONGITUDE - only 2,160 missing look at end; maybe drop
- LOCATIONonly 2,160 missing look at end; maybe drop



Confused

- WORK_ZONE_TYPE - alot missing; THERE IS UNKNOWN COLUMN; 


## Columns to drop initially (changes)

In [339]:
df.head(2)

crash_recordd     rd_no crash_date_est  \
0  00027e2894dd2f3fe4ff320a6d332d18e465b5c8ba2e79...  JC201794            NaN   
4  0006882952e53c291df267014a03b57684383e9ad66d9e...  JC459747            NaN   

               crash_date  posted_speed_limit traffic_control_device  \
0  03/27/2019 04:20:00 PM                  30            NO CONTROLS   
4  10/04/2019 12:00:00 PM                  30         TRAFFIC SIGNAL   

       device_condition weather_condition lighting_condition crash_type  \
0           NO CONTROLS             CLEAR           DAYLIGHT      ANGLE   
4  FUNCTIONING PROPERLY             CLEAR           DAYLIGHT   REAR END   

  trafficway_type  lane_cnt           alignment roadway_surface_cond  \
0     NOT DIVIDED       NaN  STRAIGHT AND LEVEL                  DRY   
4     NOT DIVIDED       NaN  STRAIGHT AND LEVEL                  DRY   

  road_defect                 report_type        injury/tow_or_no  \
0  NO DEFECTS                    ON SCENE  NO INJURY / DRIVE AWAY   
4  NO DEFECTS  NOT ON SCENE (DESK REPORT)  NO INJURY / DRIVE AWAY   

  intersection_related not_right_of_way hit_and_run         damage  \
0                  NaN              NaN         NaN  $501 - $1,500   
4                    Y              NaN         NaN    OVER $1,500   

     date_police_notified        prim_contributory_cause  \
0  03/27/2019 04:27:00 PM  FAILING TO YIELD RIGHT-OF-WAY   
4  10/04/2019 12:35:00 PM          FOLLOWING TOO CLOSELY   

       sec_contributory_cause  street_no street_direction     street_name  \
0              NOT APPLICABLE       6223                N  CALIFORNIA AVE   
4  IMPROPER TURNING/NO SIGNAL       2401                S       DAMEN AVE   

   beat_of_occurrence photos_taken statements_taken dooring work_zone  \
0              2413.0          NaN              NaN     NaN       NaN   
4              1034.0          NaN              NaN     NaN       NaN   

  work_zone_type workers_present  num_units         most_severenjury  \
0            NaN             NaN        2.0  NO INDICATION OF INJURY   
4            NaN             NaN        2.0  NO INDICATION OF INJURY   

   injuries_total  injuries_fatal  injuriesncapacitating  \
0             0.0             0.0                    0.0   
4             0.0             0.0                    0.0   

   injuries_nonncapacitating  injuries_reported_not_evident  \
0                        0.0                            0.0   
4                        0.0                            0.0   

   injuries_nondication  injuries_unknown  crash_hour  crash_day_of_week  \
0                   4.0               0.0          16                  4   
4                   2.0               0.0          12                  6   

   crash_month   latitude  longitude                                  location  
0            3  41.994704 -87.699395  POINT (-87.699395085278 41.994703544264)  
4           10  41.848490 -87.675599   POINT (-87.67559940405 41.848490427751)

In [11]:
df[["crash_date_est", "crash_date",  "crash_month","crash_day_of_week",  "crash_hour","report_type"]][:5]

crash_date_est              crash_date  crash_month  crash_day_of_week  \
0             NaN  03/27/2019 04:20:00 PM            3                  4   
4             NaN  10/04/2019 12:00:00 PM           10                  6   
6             NaN  08/17/2019 02:00:00 AM            8                  7   
10            NaN  11/09/2017 08:05:00 PM           11                  5   
12            NaN  02/10/2019 03:00:00 PM            2                  1   

    crash_hour                 report_type  
0           16                    ON SCENE  
4           12  NOT ON SCENE (DESK REPORT)  
6            2  NOT ON SCENE (DESK REPORT)  
10          20  NOT ON SCENE (DESK REPORT)  
12          15  NOT ON SCENE (DESK REPORT)

In [341]:
df["crash_date_est"].value_counts()

Y    12094
N     2114
Name: crash_date_est, dtype: int64

In [342]:
df["report_type"].value_counts()

NOT ON SCENE (DESK REPORT)    139456
ON SCENE                      106587
Name: report_type, dtype: int64

In [343]:
df[["crash_type", "injury/tow_or_no"]][:5]

crash_type                  injury/tow_or_no
0        ANGLE            NO INJURY / DRIVE AWAY
4     REAR END            NO INJURY / DRIVE AWAY
6   PEDESTRIAN  INJURY AND / OR TOW DUE TO CRASH
10    REAR END            NO INJURY / DRIVE AWAY
12    REAR END            NO INJURY / DRIVE AWAY

In [344]:
df['work_zone'].value_counts(dropna=False)

NaN    250032
Y        1656
N         401
Name: work_zone, dtype: int64

In [345]:
df['work_zone_type'].value_counts(dropna=False)

NaN             250433
CONSTRUCTION      1176
UNKNOWN            199
MAINTENANCE        175
UTILITY            106
Name: work_zone_type, dtype: int64

Drop (not helpful)

- RD_NO - alot missing; not helpful for model (unique value)
- LANE_CNT - 19,000 missing; there are really big #; WHERE DID THIS EVEN HAPPEN?
- CRASH_RECORD_ID - a unique number not helpful for regression
- crash_date_est - using report type instead (no missing values)
- photos_taken - SOO many missing
- statement_taken - SO MANY MISSING
        - should i keep or no? sec_contributory_cause
- 

In [26]:
df1 = df.drop(columns=['rd_no','lane_cnt','crash_recordd','crash_date_est','photos_taken', 'statements_taken', 'injuries_unknown', 'location'])
not_determined_df = not_determined_df.drop(columns=['rd_no','lane_cnt','crash_recordd','crash_date_est','photos_taken', 'statements_taken', 'injuries_unknown', 'location'])

df1["crash_date"] = df1["crash_date"].astype('datetime64[ns]')
df1['year'] = pd.DatetimeIndex(df1['crash_date']).year

not_determined_df["crash_date"] = not_determined_df["crash_date"].astype('datetime64[ns]')
not_determined_df['year'] = pd.DatetimeIndex(not_determined_df['crash_date']).year

In [361]:
df1.head(2)

crash_date  posted_speed_limit traffic_control_device  \
0  03/27/2019 04:20:00 PM                  30            NO CONTROLS   
4  10/04/2019 12:00:00 PM                  30         TRAFFIC SIGNAL   

       device_condition weather_condition lighting_condition crash_type  \
0           NO CONTROLS             CLEAR           DAYLIGHT      ANGLE   
4  FUNCTIONING PROPERLY             CLEAR           DAYLIGHT   REAR END   

  trafficway_type           alignment roadway_surface_cond road_defect  \
0     NOT DIVIDED  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
4     NOT DIVIDED  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   

                  report_type        injury/tow_or_no intersection_related  \
0                    ON SCENE  NO INJURY / DRIVE AWAY                  NaN   
4  NOT ON SCENE (DESK REPORT)  NO INJURY / DRIVE AWAY                    Y   

  not_right_of_way hit_and_run         damage    date_police_notified  \
0              NaN         NaN  $501 - $1,500  03/27/2019 04:27:00 PM   
4              NaN         NaN    OVER $1,500  10/04/2019 12:35:00 PM   

         prim_contributory_cause      sec_contributory_cause  street_no  \
0  FAILING TO YIELD RIGHT-OF-WAY              NOT APPLICABLE       6223   
4          FOLLOWING TOO CLOSELY  IMPROPER TURNING/NO SIGNAL       2401   

  street_direction     street_name  beat_of_occurrence dooring work_zone  \
0                N  CALIFORNIA AVE              2413.0     NaN       NaN   
4                S       DAMEN AVE              1034.0     NaN       NaN   

  work_zone_type workers_present  num_units         most_severenjury  \
0            NaN             NaN        2.0  NO INDICATION OF INJURY   
4            NaN             NaN        2.0  NO INDICATION OF INJURY   

   injuries_total  injuries_fatal  injuriesncapacitating  \
0             0.0             0.0                    0.0   
4             0.0             0.0                    0.0   

   injuries_nonncapacitating  injuries_reported_not_evident  \
0                        0.0                            0.0   
4                        0.0                            0.0   

   injuries_nondication  crash_hour  crash_day_of_week  crash_month  \
0                   4.0          16                  4            3   
4                   2.0          12                  6           10   

    latitude  longitude  
0  41.994704 -87.699395  
4  41.848490 -87.675599

## Explore new df (drop/fill in etc) (changes)

In [362]:
df1.isna().sum()

crash_date                            0
posted_speed_limit                    0
traffic_control_device                0
device_condition                      0
weather_condition                     0
lighting_condition                    0
crash_type                            0
trafficway_type                       0
alignment                             0
roadway_surface_cond                  0
road_defect                           0
report_type                        6046
injury/tow_or_no                      0
intersection_related             186332
not_right_of_way                 241081
hit_and_run                      197330
damage                                0
date_police_notified                  0
prim_contributory_cause               0
sec_contributory_cause                0
street_no                             0
street_direction                      2
street_name                           0
beat_of_occurrence                    3
dooring                          251347


In [363]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 252089 entries, 0 to 395203
Data columns (total 41 columns):
crash_date                       252089 non-null object
posted_speed_limit               252089 non-null int64
traffic_control_device           252089 non-null object
device_condition                 252089 non-null object
weather_condition                252089 non-null object
lighting_condition               252089 non-null object
crash_type                       252089 non-null object
trafficway_type                  252089 non-null object
alignment                        252089 non-null object
roadway_surface_cond             252089 non-null object
road_defect                      252089 non-null object
report_type                      246043 non-null object
injury/tow_or_no                 252089 non-null object
intersection_related             65757 non-null object
not_right_of_way                 11008 non-null object
hit_and_run                      54759 non-null obje

In [364]:
ind = df1['latitude'].isna()

In [365]:
df1.loc[ind]

crash_date  posted_speed_limit traffic_control_device  \
35      10/02/2018 06:30:00 PM                  10            NO CONTROLS   
13255   11/08/2019 01:00:00 AM                  30            NO CONTROLS   
17814   09/21/2019 07:17:00 PM                  30            NO CONTROLS   
32501   02/19/2020 10:00:00 AM                  20            NO CONTROLS   
37193   03/02/2020 08:01:00 PM                  10            NO CONTROLS   
...                        ...                 ...                    ...   
393955  07/16/2018 03:09:00 PM                  30            NO CONTROLS   
394095  12/18/2019 04:09:00 PM                  30            NO CONTROLS   
394122  06/04/2018 04:40:00 PM                  30            NO CONTROLS   
394573  04/25/2019 01:30:00 PM                  10            NO CONTROLS   
394587  01/15/2019 09:00:00 PM                  25            NO CONTROLS   

       device_condition weather_condition      lighting_condition  \
35          NO CONTROLS             CLEAR                DARKNESS   
13255       NO CONTROLS             CLEAR                DAYLIGHT   
17814       NO CONTROLS             CLEAR                DARKNESS   
32501       NO CONTROLS             CLEAR                DAYLIGHT   
37193       NO CONTROLS             CLEAR  DARKNESS, LIGHTED ROAD   
...                 ...               ...                     ...   
393955      NO CONTROLS             CLEAR                DAYLIGHT   
394095      NO CONTROLS             CLEAR                DAYLIGHT   
394122      NO CONTROLS             CLEAR                DAYLIGHT   
394573      NO CONTROLS             CLEAR                DAYLIGHT   
394587      NO CONTROLS              RAIN  DARKNESS, LIGHTED ROAD   

                      crash_type trafficway_type           alignment  \
35          PARKED MOTOR VEHICLE           OTHER  STRAIGHT AND LEVEL   
13255       PARKED MOTOR VEHICLE     NOT DIVIDED  STRAIGHT AND LEVEL   
17814               FIXED OBJECT           OTHER  STRAIGHT AND LEVEL   
32501   SIDESWIPE SAME DIRECTION         ONE-WAY  STRAIGHT AND LEVEL   
37193       PARKED MOTOR VEHICLE     PARKING LOT  STRAIGHT AND LEVEL   
...                          ...             ...                 ...   
393955      PARKED MOTOR VEHICLE         ONE-WAY  STRAIGHT AND LEVEL   
394095                  REAR END     NOT DIVIDED  STRAIGHT AND LEVEL   
394122  SIDESWIPE SAME DIRECTION         ONE-WAY  STRAIGHT AND LEVEL   
394573                PEDESTRIAN     PARKING LOT  STRAIGHT AND LEVEL   
394587  SIDESWIPE SAME DIRECTION         ONE-WAY  STRAIGHT AND LEVEL   

       roadway_surface_cond road_defect                 report_type  \
35                      DRY  NO DEFECTS                    ON SCENE   
13255                   DRY  NO DEFECTS  NOT ON SCENE (DESK REPORT)   
17814                   DRY  NO DEFECTS                    ON SCENE   
32501                   DRY  NO DEFECTS                    ON SCENE   
37193                   DRY  NO DEFECTS                    ON SCENE   
...                     ...         ...                         ...   
393955                  DRY  NO DEFECTS                    ON SCENE   
394095                  DRY  NO DEFECTS  NOT ON SCENE (DESK REPORT)   
394122                  DRY     UNKNOWN  NOT ON SCENE (DESK REPORT)   
394573                  DRY  NO DEFECTS                    ON SCENE   
394587                  WET  NO DEFECTS                    ON SCENE   

                        injury/tow_or_no intersection_related  \
35                NO INJURY / DRIVE AWAY                  NaN   
13255             NO INJURY / DRIVE AWAY                  NaN   
17814             NO INJURY / DRIVE AWAY                  NaN   
32501             NO INJURY / DRIVE AWAY                  NaN   
37193             NO INJURY / DRIVE AWAY                  NaN   
...                                  ...                  ...   
393955            NO INJURY / DRIVE AWAY                  NaN   
394095            NO INJURY 

if street_no and street_name are the same use that latitude long and location

517	W	OHARE ST	1654.0

In [366]:
df1.loc[(df1['street_no'] == 517)& (df1['street_name'] == "OHARE ST")]

crash_date  posted_speed_limit traffic_control_device  \
35      10/02/2018 06:30:00 PM                  10            NO CONTROLS   
58843   01/10/2020 04:00:00 PM                  15            NO CONTROLS   
152720  09/19/2019 04:00:00 PM                  15            NO CONTROLS   
155904  10/29/2018 09:45:00 PM                  15            NO CONTROLS   
228429  01/27/2020 02:25:00 PM                  20            NO CONTROLS   
291554  09/23/2019 09:05:00 PM                  15            NO CONTROLS   

       device_condition weather_condition      lighting_condition  \
35          NO CONTROLS             CLEAR                DARKNESS   
58843       NO CONTROLS   CLOUDY/OVERCAST                    DUSK   
152720      NO CONTROLS             CLEAR                DAYLIGHT   
155904      NO CONTROLS             CLEAR  DARKNESS, LIGHTED ROAD   
228429      NO CONTROLS             CLEAR                DAYLIGHT   
291554      NO CONTROLS             CLEAR  DARKNESS, LIGHTED ROAD   

                  crash_type                  trafficway_type  \
35      PARKED MOTOR VEHICLE                            OTHER   
58843           OTHER OBJECT                            OTHER   
152720  PARKED MOTOR VEHICLE                            OTHER   
155904  PARKED MOTOR VEHICLE                         DRIVEWAY   
228429         REAR TO FRONT  DIVIDED - W/MEDIAN (NOT RAISED)   
291554  PARKED MOTOR VEHICLE                      PARKING LOT   

                 alignment roadway_surface_cond road_defect report_type  \
35      STRAIGHT AND LEVEL                  DRY  NO DEFECTS    ON SCENE   
58843   STRAIGHT AND LEVEL                  DRY  NO DEFECTS    ON SCENE   
152720  STRAIGHT AND LEVEL                  DRY  NO DEFECTS    ON SCENE   
155904  STRAIGHT AND LEVEL                  DRY  NO DEFECTS    ON SCENE   
228429  STRAIGHT AND LEVEL                  DRY  NO DEFECTS    ON SCENE   
291554  STRAIGHT AND LEVEL                  DRY  NO DEFECTS    ON SCENE   

              injury/tow_or_no intersection_related not_right_of_way  \
35      NO INJURY / DRIVE AWAY                  NaN              NaN   
58843   NO INJURY / DRIVE AWAY                  NaN              NaN   
152720  NO INJURY / DRIVE AWAY                  NaN              NaN   
155904  NO INJURY / DRIVE AWAY                  NaN              NaN   
228429  NO INJURY / DRIVE AWAY                  NaN              NaN   
291554  NO INJURY / DRIVE AWAY                  NaN              NaN   

       hit_and_run         damage    date_police_notified  \
35             NaN    OVER $1,500  10/02/2018 07:35:00 PM   
58843          NaN  $501 - $1,500  01/10/2020 04:01:00 PM   
152720         NaN  $501 - $1,500  09/19/2019 04:08:00 PM   
155904           Y  $501 - $1,500  10/29/2018 09:53:00 PM   
228429         NaN    OVER $1,500  01/27/2020 02:30:00 PM   
291554         NaN    OVER $1,500  09/23/2019 09:10:00 PM   

                                  prim_contributory_cause  \
35                                         NOT APPLICABLE   
58843                                      NOT APPLICABLE   
152720  VISION OBSCURED (SIGNS, TREE LIMBS, BUILDINGS,...   
155904                DRIVING SKILLS/KNOWLEDGE/EXPERIENCE   
228429                                   IMPROPER BACKING   
291554                DRIVING SKILLS/KNOWLEDGE/EXPERIENCE   

       sec_contributory_cause  street_no street_direction street_name  \
35             NOT APPLICABLE        517                W    OHARE ST   
58843          NOT APPLICABLE        517                W    OHARE ST   
152720         NOT APPLICABLE        517                W    OHARE ST   
155904    UNABLE TO DETERMINE        517                W    OHARE ST   
228429    UNABLE TO DETERMINE        517                W    OHARE ST   
291554    UNABLE TO DETERMINE        517                W    OHARE ST   

        beat_of_occurrence dooring work_zone work_zone_type workers_present  \
35                  1654.0     NaN       NaN            NaN   

In [367]:
df1.loc[(df1['street_no'] == 4547)& (df1['street_name'] == "DR MARTIN LUTHER KING JR SD")]

crash_date  posted_speed_limit traffic_control_device  \
393955  07/16/2018 03:09:00 PM                  30            NO CONTROLS   

       device_condition weather_condition lighting_condition  \
393955      NO CONTROLS             CLEAR           DAYLIGHT   

                  crash_type trafficway_type           alignment  \
393955  PARKED MOTOR VEHICLE         ONE-WAY  STRAIGHT AND LEVEL   

       roadway_surface_cond road_defect report_type        injury/tow_or_no  \
393955                  DRY  NO DEFECTS    ON SCENE  NO INJURY / DRIVE AWAY   

       intersection_related not_right_of_way hit_and_run        damage  \
393955                  NaN              NaN         NaN  $500 OR LESS   

          date_police_notified prim_contributory_cause sec_contributory_cause  \
393955  07/16/2018 03:20:00 PM          NOT APPLICABLE         NOT APPLICABLE   

        street_no street_direction                  street_name  \
393955       4547                S  DR MARTIN LUTHER KING JR SD   

        beat_of_occurrence dooring work_zone work_zone_type workers_present  \
393955               221.0     NaN       NaN            NaN             NaN   

        num_units         most_severenjury  injuries_total  injuries_fatal  \
393955        2.0  NO INDICATION OF INJURY             0.0             0.0   

        injuriesncapacitating  injuries_nonncapacitating  \
393955                    0.0                        0.0   

        injuries_reported_not_evident  injuries_nondication  crash_hour  \
393955                            0.0                  10.0          15   

        crash_day_of_week  crash_month  latitude  longitude  
393955                  2            7       NaN        NaN

    Okay, so we decided to fill in and drop a few columns

In [27]:
df1['work_zone_type'].fillna("Not work zone", inplace=True)
df1['work_zone'].fillna("N", inplace=True)
df1['workers_present'].fillna("N", inplace=True)
df1.dooring.fillna("N", inplace=True)
df1['intersection_related'].fillna("N", inplace=True)
df1['not_right_of_way'].fillna("N", inplace=True)
df1['hit_and_run'].fillna("N", inplace=True)
df1= df1.dropna(subset=['report_type', 'latitude','most_severenjury', 'beat_of_occurrence'])

In [28]:
not_determined_df['work_zone_type'].fillna("Not work zone", inplace=True)
not_determined_df['work_zone'].fillna("N", inplace=True)
not_determined_df['workers_present'].fillna("N", inplace=True)
not_determined_df.dooring.fillna("N", inplace=True)
not_determined_df['intersection_related'].fillna("N", inplace=True)
not_determined_df['not_right_of_way'].fillna("N", inplace=True)
not_determined_df['hit_and_run'].fillna("N", inplace=True)
not_determined_df= not_determined_df.dropna(subset=['report_type', 'latitude','most_severenjury', 'beat_of_occurrence'])

In [29]:
df1.isna().sum()

crash_date                       0
posted_speed_limit               0
traffic_control_device           0
device_condition                 0
weather_condition                0
lighting_condition               0
crash_type                       0
trafficway_type                  0
alignment                        0
roadway_surface_cond             0
road_defect                      0
report_type                      0
injury/tow_or_no                 0
intersection_related             0
not_right_of_way                 0
hit_and_run                      0
damage                           0
date_police_notified             0
prim_contributory_cause          0
sec_contributory_cause           0
street_no                        0
street_direction                 0
street_name                      0
beat_of_occurrence               0
dooring                          0
work_zone                        0
work_zone_type                   0
workers_present                  0
num_units           

In [370]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 243583 entries, 0 to 395203
Data columns (total 41 columns):
crash_date                       243583 non-null object
posted_speed_limit               243583 non-null int64
traffic_control_device           243583 non-null object
device_condition                 243583 non-null object
weather_condition                243583 non-null object
lighting_condition               243583 non-null object
crash_type                       243583 non-null object
trafficway_type                  243583 non-null object
alignment                        243583 non-null object
roadway_surface_cond             243583 non-null object
road_defect                      243583 non-null object
report_type                      243583 non-null object
injury/tow_or_no                 243583 non-null object
intersection_related             243583 non-null object
not_right_of_way                 243583 non-null object
hit_and_run                      243583 non-null o

## Look into Place holders (questions in here!!!)

In [30]:
for col in df1.columns:
    print(col,'\n', df1[col].value_counts(dropna=False), "\n\n")

crash_date 
 2017-11-10 10:30:00    17
2019-01-12 14:00:00    13
2017-11-10 10:00:00    12
2020-02-26 08:00:00    12
2019-01-12 15:00:00    11
                       ..
2017-01-11 19:00:00     1
2020-02-18 11:49:00     1
2019-11-15 19:30:00     1
2017-10-10 19:40:00     1
2017-12-31 22:00:00     1
Name: crash_date, Length: 160280, dtype: int64 


posted_speed_limit 
 30    155748
35     15010
25     11251
20      6521
15      6250
10      2989
0       2912
40      2287
5       1466
45      1396
55       161
3         52
50        39
9         38
39        20
2         12
60        10
99         7
32         5
1          5
11         4
70         2
65         2
7          2
6          2
33         2
34         2
22         1
24         1
12         1
4          1
38         1
63         1
Name: posted_speed_limit, dtype: int64 


traffic_control_device 
 NO CONTROLS                 112394
TRAFFIC SIGNAL               66601
STOP SIGN/FLASHER            23707
OTHER                        

beat_of_occurrence 
 1834.0    3094
114.0     2870
122.0     2690
1831.0    2305
813.0     2242
          ... 
231.0      224
232.0      216
422.0      169
1125.0     162
1654.0     141
Name: beat_of_occurrence, Length: 271, dtype: int64 


dooring 
 N    205789
Y       412
Name: dooring, dtype: int64 


work_zone 
 N    204985
Y      1216
Name: work_zone, dtype: int64 


work_zone_type 
 Not work zone    204985
CONSTRUCTION        964
MAINTENANCE         157
UTILITY              95
Name: work_zone_type, dtype: int64 


workers_present 
 N    205836
Y       365
Name: workers_present, dtype: int64 


num_units 
 2.0     181792
3.0      11162
1.0      10651
4.0       1945
5.0        451
6.0        118
7.0         43
8.0         18
9.0          8
10.0         7
12.0         2
11.0         2
15.0         1
18.0         1
Name: num_units, dtype: int64 


most_severenjury 
 NO INDICATION OF INJURY     176339
NONINCAPACITATING INJURY     16536
REPORTED, NOT EVIDENT         9367
INCAPACITATING

posted_speed_limit 
    - why is there 99 or 70 or 4?
    - Posted speed limit, as determined by reporting office
    - WHAT THE HELL DO I DO WITH OTHER OR UNKNOW???
    
sec_contributory_cause
    - What do i do with not applicable and unable to determine?
    
    
- if a column is Y or N - can i just keep it like that? no dummifying right?

## Data type

In [32]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 206201 entries, 0 to 395203
Data columns (total 42 columns):
crash_date                       206201 non-null datetime64[ns]
posted_speed_limit               206201 non-null int64
traffic_control_device           206201 non-null object
device_condition                 206201 non-null object
weather_condition                206201 non-null object
lighting_condition               206201 non-null object
crash_type                       206201 non-null object
trafficway_type                  206201 non-null object
alignment                        206201 non-null object
roadway_surface_cond             206201 non-null object
road_defect                      206201 non-null object
report_type                      206201 non-null object
injury/tow_or_no                 206201 non-null object
intersection_related             206201 non-null object
not_right_of_way                 206201 non-null object
hit_and_run                      206201 no

# Saved files

In [31]:
df1['target_names'] = df1['prim_contributory_cause']
df1.drop('prim_contributory_cause', axis=1, inplace=True)

In [182]:
df1.to_csv("clean_car_crash.csv", index=False)

In [183]:
not_determined_df.to_csv("not_determined.csv", index=False)

Drop

- work_zone_type
- street_dierction
- MAYBE sec_contributory_cause
- MAYBE street_name
- date_police_notified


- road_defect - drop unknown
- road_surface_condition - drop unknown intersection type
- trafficway_type - drop unknown
- lighting_condition - drop unknown
- weather_condition - drop unknown
- device consition - drop unknown/missing
- traffic_control_device - drop unknown

# Other


In [191]:
for num in list(range(2,102, 5)):
    print(num)

2
7
12
17
22
27
32
37
42
47
52
57
62
67
72
77
82
87
92
97
